### PR TITLE
Implementation to support JSON music descriptions

### PR DESCRIPTION
**Pull request by OrenAudeles**
#### Summary
```SUMMARY: Content "Implementation to support use of JSON snippets for procedural music descriptions"```

#### Purpose of change
```Implementation was required to support the use of JSON snippets when generating music descriptions provided in PR``` #25728

#### Describe the solution
Pretty easy implementation. Given the snippet categories in the linked PR's JSON file, pick randomly from each category and stitch the returned values together. In the case there are no entries in any category the result of random picking will be an empty string, and cause the music description to be returned empty. This is to allow removal of music genre snippets entirely without causing errors. The return type of the function had to be changed from string by-ref to string by-val so that we do not return a reference to a temporary value. Genre snippet categories were placed in an array for ease of use.

On average 24.5% chance of returning a JSON based music description.

### CODE REVIEW

File: src/iuse.cpp [hunk 1]
```diff
@@ -3590,7 +3591,7 @@ int iuse::mp3( player *p, item *it, bool, const tripoint & )
     return it->type->charges_to_use();
 }
 
-const std::string &get_music_description()
+const std::string get_music_description()
```
Comment (BevapDin) [inline src/iuse.cpp]: Please don't return a const value. See https://stackoverflow.com/questions/6299967 for why it's pointless. If you don't want to return a reference anymore, return a non-const value.

File: src/iuse.cpp [hunk 2]
```diff
@@ -3603,6 +3604,11 @@ const std::string &get_music_description()
 
         }
     };
+    static const std::array<std::string, 3> genre_snippet_categories = {{
```
Comment (BevapDin) [inline src/iuse.cpp]: This array is quite useless. You access each element exactly once and with a fixed index, not using any loop or similar. You can use the string literals directly in the function calls below. And why is this defined so far away from its actual usage?

File: src/iuse.cpp [hunk 3]
```diff
@@ -3612,6 +3618,19 @@ const std::string &get_music_description()
...
+    // of snippets {a, b, c}, but only a 50% chance
+    // Actual chance = 24.5% of being selected
+    if( one_in( 2 ) ) {
+        const std::string& from_a = SNIPPET.random_from_category( genre_snippet_categories[0] );
+        const std::string& from_b = SNIPPET.random_from_category( genre_snippet_categories[1] );
+        const std::string& from_c = SNIPPET.random_from_category( genre_snippet_categories[2] );
+
+        // Require all to be non-empty
+        if( !( from_a.empty() | from_b.empty() | from_c.empty() ) ) {
+            return string_format( "%s%s%s", from_a.c_str(), from_b.c_str(), from_c.c_str() );
```
Comment (BevapDin) [inline src/iuse.cpp]:
```suggestion
            return from_a + from_b + from_c;
```

File: src/iuse.cpp [hunk 4]
```diff
@@ -3612,6 +3618,19 @@ const std::string &get_music_description()
...
+    // Not one of the hard-coded versions, let's apply a random string made up
+    // of snippets {a, b, c}, but only a 50% chance
+    // Actual chance = 24.5% of being selected
+    if( one_in( 2 ) ) {
+        const std::string& from_a = SNIPPET.random_from_category( genre_snippet_categories[0] );
+        const std::string& from_b = SNIPPET.random_from_category( genre_snippet_categories[1] );
+        const std::string& from_c = SNIPPET.random_from_category( genre_snippet_categories[2] );
+
+        // Require all to be non-empty
+        if( !( from_a.empty() | from_b.empty() | from_c.empty() ) ) {
```
Comment (BevapDin) [inline src/iuse.cpp]:
```suggestion
        if( !( from_a.empty() || from_b.empty() || from_c.empty() ) ) {
```

**Comment by BevapDin** [comment]
It also needs some examples because currently it's just dead code.

File: src/iuse.cpp (hunk 1 above)
Comment (OrenAudeles) [reply src/iuse.cpp]: Good point on not returning a const value, thanks.

**Comment by OrenAudeles** [comment]
Example JSON file is in #25728
Example screenshot:
screenshot from 2018-11-10 17-17-48

**Pull request merged**